src/cuda/CUDACore/allocate_device.cc

```diff
@@ -14,22 +14,28 @@ namespace {
 namespace cms::cuda {
   void *allocate_device(int dev, size_t nbytes, cudaStream_t stream) {
     void *ptr = nullptr;
-    if constexpr (allocator::useCaching) {
+    if constexpr (allocator::policy == allocator::Policy::Caching) {
       if (nbytes > maxAllocationSize) {
         throw std::runtime_error("Tried to allocate " + std::to_string(nbytes) +
                                  " bytes, but the allocator maximum is " + std::to_string(maxAllocationSize));
       }
       cudaCheck(allocator::getCachingDeviceAllocator().DeviceAllocate(dev, &ptr, nbytes, stream));
+    } else if constexpr (allocator::policy == allocator::Policy::Asynchronous) {
+      ScopedSetDevice setDeviceForThisScope(dev);
+      cudaCheck(cudaMallocAsync(&ptr, nbytes, stream));
```

Comment:
This fails to compile on CUDA 11.1, so needs to be protected with `#ifdef` as well (or the implementation moved to a templated function to allow `if constexpr` to avoid checking the discarded statement).

Reply:
I don't understand - this *is* within an  `if constexpr` that should not be taken if CUDA < 11.2

Reply:
Right, but
> Outside a template, a discarded statement is fully checked. `if constexpr` is not a substitute for the `#if` preprocessing directive

https://en.cppreference.com/w/cpp/language/if

Reply:
Sigh.

Reply:
Yeah...

Reply:
So, what do you prefer ?

1. add a declaration for the new functions if the CUDA version is older than 11.2.0:
    ```c++
    #include <cuda_runtime.h>
    #if CUDA_VERSION < 11020
   // declare the stream ordered memory allocation functions, that are available only starting from CUDA 11.2.0
    __host__ cudaError_t cudaFreeAsync(void* ptr, cudaStream_t stream);
    __host__ cudaError_t cudaMallocAsync(void** ptr, size_t size, cudaStream_t stream);
    #endif  // CUDA_VERSION
    ```

2. add an `#if CUDA_VERSION >= 11020` around the use of `cudaMallocAsync`/`cudaFreeAsync`

3. add some fake templating to avoid checking inside the `if constexpr`

Reply:
I don't have strong feeling. I'd probably go myself with 2, but I'd be fine with any of them (for the purposes of this repo).

Reply:
I've added some `#if` ...